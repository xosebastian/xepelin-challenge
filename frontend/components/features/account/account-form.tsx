"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import * as z from "zod";

import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { useForm } from "react-hook-form";

export const formAccountSchema = z.object({
  accountName: z.string().min(2, {
    message: "The name should have at least 2 characters.",
  }),
  accountNumber: z
    .string()
    .min(10, {
      message: "The account number should have at least 10 digits.",
    })
    .regex(/^\d+$/, {
      message: "The account number should only contain numbers.",
    }),
  balance: z
    .string()
    .max(5, {
      message: "The balance should have at most 5 digits.",
    })
    .regex(/^\d+(\.\d{1,2})?$/, {
      message: "The balance should be a number with up to 2 decimal places.",
    })
    .transform((value) => parseFloat(value)),
});

interface Props {
  onSubmit: (values: z.infer<typeof formAccountSchema>) => void;
  onChange?: () => void;
}

export function AccountForm({ onSubmit, onChange }: Props) {
  const form = useForm<z.infer<typeof formAccountSchema>>({
    resolver: zodResolver(formAccountSchema),
    defaultValues: {
      accountName: "",
      accountNumber: "",
      balance: 0,
    },
  });

  return (
    <Form {...form}>
      <form
        onSubmit={form.handleSubmit(onSubmit)}
        className="space-y-2 p-5"
        onChange={onChange}
      >
        <FormField
          control={form.control}
          name="accountName"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Name</FormLabel>
              <FormControl>
                <Input placeholder="Enter your account name" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="accountNumber"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Number</FormLabel>
              <FormControl>
                <Input
                  maxLength={10}
                  placeholder="Enter your account number"
                  {...field}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="balance"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Balance</FormLabel>
              <FormControl>
                <Input type="number" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <Button role="button" type="submit">
          Create
        </Button>
      </form>
    </Form>
  );
}
