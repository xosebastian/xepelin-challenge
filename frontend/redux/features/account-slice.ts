import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'

export interface AccountState {
  id: string,
  balance: number,
}

const initialState: AccountState = {
  id: '',
  balance: 0,
}

export const accountSlice = createSlice({
  name: 'account',
  initialState,
  reducers: {
    create: (state, action: PayloadAction<{id : string}>) => {
      state.id = action.payload.id
    },
    clear: () => {
      return initialState
    },
    update: (state, action: PayloadAction<{balance: number}>) => {
      state.balance = action.payload.balance
    },
    transactions: (state, action: PayloadAction<{balance: number, type: string}>) => {
      if (action.payload.type === 'DEPOSIT') {
        state.balance += action.payload.balance
      }else if (action.payload.type === 'WITHDRAWAL') {
        state.balance -= action.payload.balance
      }
    },
  },
})


export const { create, clear, update, transactions } = accountSlice.actions

export default accountSlice.reducer