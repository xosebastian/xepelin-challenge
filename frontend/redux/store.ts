import { configureStore } from '@reduxjs/toolkit'
import accountReducer from './features/account-slice'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'

export const store = configureStore({
  reducer: {
    account: accountReducer,
  },
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;