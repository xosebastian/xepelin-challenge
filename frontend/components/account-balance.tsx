"use client";
import { useAppSelector } from "@/redux/store"

export function AccountBalance () {

  const balance = useAppSelector((state) => state.account.balance);

  return (
    <div className="flex flex-col space-y-1.5 p-6">
      <h3 className="text-2xl font-semibold leading-none tracking-tight">
        $ {balance}
      </h3>
    </div>
  )
}