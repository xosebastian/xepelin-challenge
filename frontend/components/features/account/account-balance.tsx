"use client";
import { useAppDispatch, useAppSelector } from "@/redux/store";
import axios, { AxiosResponse } from "axios";
import { useEffect } from "react";
import { update } from "@/redux/features/account-slice";

export function AccountBalance() {
  const accountId = useAppSelector((state) => state.account.id);
  const balance = useAppSelector((state) => state.account.balance);
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (!accountId) return;

    const getBalance = async () => {
      try {
        const {
          data: { balance },
        }: AxiosResponse<{ balance: string }> = await axios.get(
          `/api/account/balance/${accountId}`,
          {}
        );

        console.log(balance);

        dispatch(
          update({
            balance: parseFloat(balance),
          })
        );
      } catch (error) {
        console.error(error);
      }
    };

    getBalance();
  }, [accountId, dispatch]);

  return (
    <div className="flex flex-col space-y-1.5 p-6">
      <h3 className="text-2xl font-semibold leading-none tracking-tight" data-testid="balance">
        $ {balance}
      </h3>
    </div>
  );
}
