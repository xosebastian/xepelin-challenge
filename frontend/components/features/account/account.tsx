"use client";

import * as z from "zod";

import { Button } from "@/components/ui/button";

import { useAppDispatch, useAppSelector } from "@/redux/store";
import { create, clear } from "@/redux/features/account-slice";
import axios, { AxiosResponse, isAxiosError } from "axios";
import { useState } from "react";
import { Alert, AlertDescription, AlertTitle } from "../../ui/alert";
import { useToast } from "../../ui/use-toast";
import { AccountForm, formAccountSchema } from "./account-form";

export function Account() {
  const dispatch = useAppDispatch();
  const accountId = useAppSelector((state) => state.account.id);
  const [error, setError] = useState<string | null>(null);
  const { toast } = useToast();

  async function onSubmit(values: z.infer<typeof formAccountSchema>) {
    const { accountName, accountNumber, balance } = values;

    try {
      const {
        data: { accountId },
      }: AxiosResponse<{ accountId: string }> = await axios.post(
        "/api/account/create",
        {
          accountName,
          accountNumber,
          balance,
        }
      );

      dispatch(
        create({
          id: accountId,
        })
      );

      toast({
        title: "Account created",
        description: `Account ${accountName} created with id ${accountId}`,
      });
    } catch (error: any) {
      const message = isAxiosError(error)
        ? error?.response?.data.message
        : error?.message ?? "Something went wrong";

      setError(message);
    }
  }

  return (
    <>
      {accountId ? (
        <div className="flex flex-col space-y-1.5 p-6">
          <h3 className="text-2xl font-semibold leading-none tracking-tight">
            My Account
          </h3>
          <p className="text-sm text-muted-foreground">id: {accountId}</p>
          <Button
            onClick={() => {
              dispatch(clear());
            }}
          >
            Create Another
          </Button>
        </div>
      ) : (
        <>
          <div className="flex flex-col space-y-1.5 p-6">
            <h3 className="text-2xl font-semibold leading-none tracking-tight">
              Create Account
            </h3>
          </div>
          <AccountForm onSubmit={onSubmit} onChange={() => setError(null)} />
          {error && (
            <Alert variant={"destructive"} className="mt-5">
              <AlertTitle>Ups!</AlertTitle>
              <AlertDescription>{error}</AlertDescription>
            </Alert>
          )}
        </>
      )}
    </>
  );
}
