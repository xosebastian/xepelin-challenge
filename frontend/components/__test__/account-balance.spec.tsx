import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "@/redux/store";
import { AccountBalance } from "../features/account/account-balance";
import { create, update, clear } from "@/redux/features/account-slice";


describe("AccountBalance", () => {

  beforeEach(() => {
    store.dispatch(clear());
  });

  it("should render the account balance when the balance is defined", () => {
    const balance = 100.0;

    store.dispatch(create({
      id: "user-id"
    }));
 
    store.dispatch(update({
      balance
    }));

    const { container } = render(
      <Provider store={store}>
        <AccountBalance/>
      </Provider>
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByTestId('balance').textContent).toEqual(`$ ${balance}`);
  });

  it("should render an error message when the balance is not defined", () => {
    

    

    const { container } = render(
      <Provider store={store}>
        <AccountBalance  />
      </Provider>
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByTestId('balance').textContent).toEqual(`$ 0`);
  });
});