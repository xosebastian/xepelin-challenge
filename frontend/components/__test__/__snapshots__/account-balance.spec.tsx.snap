// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountBalance should render the account balance when the balance is defined 1`] = `
<div>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="text-2xl font-semibold leading-none tracking-tight"
      data-testid="balance"
    >
      $ 
      100
    </h3>
  </div>
</div>
`;
